src/hash_value.cpp

```diff
@@ -43,7 +47,11 @@ ValuePtr HashValue::set_default(Env *env, ValuePtr value) {
 void HashValue::put(Env *env, ValuePtr key, ValuePtr val) {
     assert_not_frozen(env);
     Key key_container;
+    if (key->is_string() && !key->is_frozen()) {
+        key = key->as_string()->dup(env);
+    }
```

Comment:
I had no idea Ruby did this. Seems so inefficient!

Here's something else that's weirding me out:

```ruby
s = "foo"
s2 = s.dup

h1 = { s => "bar" }
h2 = { s => "bar" }
h3 = { s2 => "bar" }
h4 = { "foo" => "bar" }

p s.object_id             # => 60
p h1.keys.first.object_id # => 80
p h2.keys.first.object_id # => 80
p h3.keys.first.object_id # => 80
p h4.keys.first.object_id # => 80
```

It's as if there is a cache of these `dup`d string keys and MRI somehow reuses them. :exploding_head:

Reply:
Yeah that was a bit weird to me too. I just added it because the specs demanded for it 😅